### PR TITLE
[Feature] Added back off if dataplane returns 451 #8

### PR DESCRIPTION
**Pull request by pallabmaiti**
## Description of the change

> Added back off in code flow. If the dataplane returns 451, there is no back-off.

## Type of change
- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: analytics.go [hunk 1]
```diff
@@ -418,12 +418,19 @@ func (c *client) send(msgs []message) {
...
 				*/
-				c.logf("Retrying in 5 seconds")
-				time.Sleep(5 * time.Second)
-				c.setNodeCount()
-				newMsgs := c.getRevisedMsgs(nodePayload, k)
-				c.send(newMsgs)
-				return
+				var sleepTimeOut = time.Duration(count*5) * time.Second
+				if sleepTimeOut.Seconds() > 300 {
+					c.debugf("Discarding events")
```
Comment (atzoum) [inline analytics.go:423]: We don't want to discard the event, just keep retrying until it succeeds. Using a maximum value for the timeout is desirable though.

File: analytics.go [hunk 2]
```diff
@@ -387,7 +387,7 @@ func (c *client) getMarshalled(msgs []message) ([]byte, error) {
 }
 
 // Send batch request.
-func (c *client) send(msgs []message) {
+func (c *client) send(msgs []message, count int) {
```
Comment (atzoum) [inline analytics.go:390]: attempt is more appropriate?

```suggestion
func (c *client) send(msgs []message, attempt int) {
```

Comment (atzoum) [inline analytics.go:390]: can we add some test cases covering this scenario?

File: analytics.go [hunk 3]
```diff
@@ -418,12 +418,19 @@ func (c *client) send(msgs []message) {
 				We would then reset the node count by making a call to configure-info end point, then regenerate the payload at a node level
 				for only those nodes where we failed in sending the data and then recursively call the send function with the updated payload.
 				*/
-				c.logf("Retrying in 5 seconds")
-				time.Sleep(5 * time.Second)
-				c.setNodeCount()
-				newMsgs := c.getRevisedMsgs(nodePayload, k)
-				c.send(newMsgs)
-				return
+				var sleepTimeOut = time.Duration(count*5) * time.Second
```
Comment (atzoum) [inline analytics.go:421]: backoff should be enforced only after the 2nd attempt

File: analytics.go (hunk 1 above)
Comment (pallabmaiti) [reply analytics.go:423]: @SumanthPuram mentioned keeping the maximum time to 5 minutes.

File: analytics.go (hunk 2 above)
Comment (pallabmaiti) [reply analytics.go:390]: Will change this.

Comment (pallabmaiti) [reply analytics.go:390]: Can we add test case later? This is a very urgent fix & the customer is waiting for it.

File: analytics.go (hunk 1 above)
Comment (atzoum) [reply analytics.go:423]: we are guarding against a future bug in proxy. Under normal conditions the proxy won't keep returning a 451. So, 5 minutes is not enough to pinpoint and fix a production issue and will lead to data loss. If we keep the max backoff timeout to 5 minutes, that's ok, but we need to keep retrying

Comment (pallabmaiti) [reply analytics.go:423]: So, what you suggest here?

Comment (pallabmaiti) [reply analytics.go:423]: Got it.

File: analytics.go [hunk 4]
```diff
@@ -418,9 +418,17 @@ func (c *client) send(msgs []message) {
 				We would then reset the node count by making a call to configure-info end point, then regenerate the payload at a node level
 				for only those nodes where we failed in sending the data and then recursively call the send function with the updated payload.
 				*/
+				var sleepTimeOut = time.Duration((retryAttempt-1)*5) * time.Second
+				if retryAttempt > 1 {
```
Comment (atzoum) [inline analytics.go:422]:
```suggestion
				if sleepTimeOut > 0 {
```

File: analytics.go [hunk 5]
```diff
@@ -418,9 +418,17 @@ func (c *client) send(msgs []message) {
 				We would then reset the node count by making a call to configure-info end point, then regenerate the payload at a node level
 				for only those nodes where we failed in sending the data and then recursively call the send function with the updated payload.
 				*/
+				var sleepTimeOut = time.Duration((retryAttempt-1)*5) * time.Second
```
Comment (atzoum) [inline analytics.go:421]:
```suggestion
				const maxSleepTime = 300*time.Second
				var sleepTimeOut = time.Duration((retryAttempt)*5) * time.Second
				if sleepTimeOut > maxSleepTime {
					sleepTimeOut = maxSleepTime
				}
```

File: analytics.go [hunk 6]
```diff
@@ -418,9 +418,17 @@ func (c *client) send(msgs []message) {
...
+				if retryAttempt > 1 {
+					c.debugf("Retrying in %d seconds", int(sleepTimeOut.Seconds()))
+					time.Sleep(sleepTimeOut)
+				}
 				c.setNodeCount()
 				newMsgs := c.getRevisedMsgs(nodePayload, k)
-				c.send(newMsgs)
+				if sleepTimeOut.Seconds() < 300 {
+					retryAttempt += 1
+				}
```
Comment (atzoum) [inline analytics.go:430]:
```suggestion
				retryAttempt += 1
```

File: analytics.go [hunk 7]
```diff
@@ -300,7 +300,7 @@ func (c *client) sendAsync(msgs []message, wg *sync.WaitGroup, ex *executor) {
 				c.errorf("panic - %s", err)
 			}
 		}()
-		c.send(msgs)
+		c.send(msgs, 1)
```
Comment (atzoum) [inline analytics.go:303]:
```suggestion
		c.send(msgs, 0)
```

File: .vscode/launch.json [hunk 8]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (atzoum) [inline .vscode/launch.json:1]: should we remove this?

**Review by atzoum — CHANGES_REQUESTED** [review]

**Review by lvrach — APPROVED** [review]

File: .vscode/launch.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (arnab-p) [reply .vscode/launch.json:1]: Let's put .vscode in .gitignore

**Review by atzoum — APPROVED** [review]

**Pull request merged**